### ISSUE TITLE
Proposal: Version number should be specified on dependencies and not in packages

### ISSUE THREAD

**Issue by judofyr** [ziglang/zig#19765]
TL;DR:

Instead of this:

```zig
// A library
.{
    .name = "hello",
    .version = "1.2.8",

    .dependencies = .{},
}

// A user of said library
.{
    .name = "world",
    .version = "2.0.0",

    .dependencies = .{
        .hello = .{
            .url = "https://example.com/v1.2.8.tar.gz",
            .hash = "...",
        },
    },
}
```

… we _remove_  the `name` and `version` field from this file and instead require the dependents to specify the version:

```zig
// A library
.{
    .dependencies = .{},
}

// A user of said library
.{
    .dependencies = .{
        .hello = .{
            .url = "https://example.com/v$version.tar.gz",
            .version = "1.2.8",
            .hash = "...",
        },
    },
}
```

## Premise

Currently, as far as I understand, a Zig package is identified by it's **name** and **version** which is declared in its `build.zig.zon`. I think this is problematic in a decentralized world and here's a proposal which completely gets rid of those.

How Zig should handle version selection has been discussed and approved in https://github.com/ziglang/zig/issues/8284#issuecomment-1000565560, but this appears to be focused around _how_ a version is selected. This is a proposal which touches on the whole concept of "what even is a package?" so I felt it deserved its own proposal. I was also a bit unsure if this is considered a "Language proposal" or not. I don't mind if you close this quickly because it doesn't belong here or if I've misunderstood something and the premise is wrong.

### Problem: A static `version` field in the source code is most likely wrong

Let's imagine the following development process:

* Commit A: Version 1.2.8 is released! `.version` is updated and the commit is tagged. **Good:** The `.version` represents the actual release.
* Commit B: A bug fix lands. **Confusing:** `.version` says 1.2.8, but in reality this is a development version.
* Commit C: A breaking change lands. **Breaking:** Now `.version` says 1.2.8, but it's actually not even compatible with the existing 1.2.8 release.
* Commit D: The `.version` is bumped to 2.0.0 to _prepare_ for the new release, but nothing is tagged. **Confusing:** `.version` says 2.0.0, but that has never even been released.
* Commit E: Version 2.0.0 is released! The commit is tagged. **Good:** The `.version` represents the actual release.

_Any_ of these commit's tree might exist in a Zig dependency tree. The most common case will be to depend on explicitly released versions, but (1) we can't actually ever enforce that and (2) if you're developing an application then it's quite convenient to be able to refer to un-released versions (e.g. pointing a library to a fork while you're waiting for the patch to accepted upstream).

This means that it's hard for the Zig build system to make any safe assumptions around which packages should be considered "the same" or "compatible".

### Problem: We can't depend on the user "doing the right thing"

It's easy to make rules such as "oh, but the developer _should_ set the version field to `.version = "1.2.3-pre` right after a release", but I'd argue that this is bound to not work in practice. Users _are_ going to forget about this. Users _are_ going to fork repos. Other users _will_ then depend on said fork in their publicly released package. Zig should be able to _detect_ when this happens and do something understandable (which could potentially be "report an error"). It should not however silently make assumption which confuses the user (i.e. your dependency suddenly gets replaced by a complete different package because it shares a name with a new dependency deeply down the tree).

### Problem: `name` is unreliable in a decentralized world

Maybe I misunderstood something from the discussion in #8284, but it appeared that the identification of a "package" would be based on the `name` it declares itself. In a centralized system it's trivial to ensure that this is globally unique, but we can't make such assumption in the decentralized world.

## Proposal: `version` on a dependency + packages are identified by their URL

Step 1: There's now two ways of declaring a dependency:

* **Without `version`:** This is the way of signaling "I want _this_ specific tarball and nothing else". You're guaranteed to get this code. This is actually what Zig already implements today. We would then keep this functionality.
* **With `version`:** This is a way of signaling "I want something that satisfies this (semantic) version, and btw here's the current tarball as well".

Step 2: Introduce `$version` parameter inside the URL (and then require `$$` as an escape sequence for writing a plain dollar).

Step 3: A package is identified by its `.url` (before expansion of `$version`). If the same URL  appears in a Zig dependency tree then it's the considered the same package. The logic specified in https://github.com/ziglang/zig/issues/8284#issuecomment-1000565560 then applies: We can pick the maximum version and use it across the whole build. We can also allow multiple major versions if we want.

This means that if package A and package B depends on different versions of package C _without_ a `version` field, then they actually depend on "different" packages from Zig's perspective. If a package is consuming or returning another package types you _definitely_ want to opt in to semantic versioning. If you only internally depend on a package you can opt in to semantic versioning if you want to play nice with the community _OR_ you can opt out if it's important for you that you get _this_ exact tarball. In the last case the user will _not_ get any version conflicts since the dependency is considered it's "unique" package.

### Typical workflow on GitHub

* A user adds `.url = "https://github.com/OWNER/REPO/archive/v$version.tar.gz", .version = "X.Y.Z"` as a dependency
* The only thing a library author needs to do for a release is `git tag vX.Y.Z && git push --tags`.
